src/search/merge_and_shrink/transition_system.cc

```diff
@@ -328,108 +327,114 @@ void TransitionSystem::apply_abstraction(
+      label is "non-empty" in the sense that there are still labels
+      associated to it.
 
       The other case is more involved: again remove all old labels from their
-      groups, and the groups themselves if they become empty. Also collect
-      the transitions of all reduced labels. Add a new group for every new
-      label and assign the collected transitions to this group. Recompute the
-      cost of all groups and compute locally equivalent labels.
-
-      NOTE: Previously, this latter case was computed in a more incremental
-      fashion: Rather than recomputing cost of all groups, we only recomputed
-      cost for groups from which we actually removed labels (hence temporarily
-      storing these affected groups). Furthermore, rather than computing
-      locally equivalent labels from scratch, we did not per default add a new
-      group for every label, but checked for an existing equivalent label
-      group. In issue539, it turned out that this incremental fashion of
-      computation does not accelerate the computation.
+      local labels, and the local labels themselves if they do not represent
+      any labels anymore. Also collect the transitions of all reduced
+      labels. Add a new local label for every new label and assign the
+      collected transitions to it. Recompute the cost of all affected local
+      labels and re-compute locally equivalent labels.
     */
 
     if (only_equivalent_labels) {
-        label_equivalence_relation->apply_label_mapping(label_mapping);
+        // Update both label mappings.
+        for (const pair<int, vector<int>> &mapping : label_mapping) {
+            int new_label = mapping.first;
+            const vector<int> &old_labels = mapping.second;
+            assert(old_labels.size() >= 2);
+            int local_label = label_to_local_label[old_labels.front()];
+            LabelGroup &label_group = local_label_infos[local_label].label_group;
+            label_group.push_back(new_label);
+            label_to_local_label[new_label] = local_label;
+
+            for (int old_label : old_labels) {
+                assert(label_to_local_label[old_label] == local_label);
+                label_group.erase(find(label_group.begin(), label_group.end(), old_label));
+                // Reset (for consistency only, old labels are never accessed).
+                label_to_local_label[old_label] = -1;
+                // NOTE: if we were combining labels with different cost,
+                // we would need to recompute the cost of the local label here.
+            }
+        }
     } else {
         /*
-          Go over all mappings, collect transitions of old groups and
-          remember all affected group IDs. This needs to happen *before*
-          updating label_equivalence_relation, because after updating it,
-          we cannot find out the group ID of reduced labels anymore.
+          Go over all label reductions (given by label_mapping). For each
```

Comment:
I'm not a great fan of the terminology here ("all label reductions"). In the terminology of our papers, the whole label mapping defines *one* label reduction transformation. But if you prefer it this way, no objections.